### DISCUSSION TITLE
In loop trying to get DHCP configuration

### DISCUSSION THREAD

**Discussion by jorgeventura** [ipxe/ipxe#745] [General]
I am trying to create an small EC2 instance but the system stays in loop at boot.
I tried to stop with user script "shell" but of course doesn't work because the loop is before my user script. 

![20220915_153759](https://user-images.githubusercontent.com/85666/190519614-3cc929a0-63a3-4ce5-bc8f-bcc59127582b.jpg)

Ventura

**Comment by NiKiZe** [comment]
So you are having a configuration issue?
There is some limitations on these instances, also see recent posts. Maybe there is answers before opening a new issue.

**Reply by jorgeventura** [reply]
I created this instance in the same VPC and Subnet from another instance (using Debian arm64 ami) and this one using iPXE ami. The debian instance has no problem to acquire IP using DHCP but the ipxe instance fail all the time and get in loop. I think the embedded script should stop at shell prompt instead of keep in loop.
I am using arm64; I think I have to try using x86_64 to verify if there is a problem with ipxe's dhcp for arm64 processors.

**Comment by jorgeventura** [comment]
How can I create my own ami for iPXE ? Is there any instruction how to create EBC EXC2 from scratch ?

**Reply by jorgeventura** [reply]
https://docs.aws.amazon.com/snowball/latest/developer-guide/ec2-ami-import-cli.html

**Reply by jorgeventura** [reply]
There is also instructions that I didn't see in the ipxe site.

**Reply by NiKiZe** [reply]
The git commits when ec/ami was added has some god documentstion

**Reply by jorgeventura** [reply]
The compiler in use is "aarch64-unknown-linux-gnu-gcc (Gentoo 12.2.0 p1) 12.2.0" and to compile with no error three files had to be patched to suppress some errors that I suspect it's caused compiler version.

These files are:

1. src/arch/arm/include/ipxe/arm_io.h (#pragma GCC diagnostic ignored "-Warray-bounds")
2. src/crypto/deflate.c (#pragma GCC diagnostic ignored "-Wdangling-pointer")
3. src/net/fc.c (#pragma GCC diagnostic ignored "-Warray-bounds")

Hopefully the patch doesn't cause any problem.

**Comment by jorgeventura** [comment]
I created an ami with a change in the aws.ipxe to drop into shell if dhcp failed.
Now I can see that there is no network card. The pci folder is empty.
I have no idea what try at this point. Any suggestion is welcome.

![screen](https://user-images.githubusercontent.com/85666/190878287-7b58c0a0-8ae9-4f15-92b8-c35ee0aec872.png)

**Reply by stappersg** [reply]
> Any suggestion is welcome.

Contact your hardware vendor and ask for which network card you should be looking for.
Yes, I'm aware that the hardware could be virtual.
We are all aware that the vendor is real.

**Comment by jorgeventura** [comment]
This is the information I have. The kernel here is not from aws; it's one generated from mainline and it's working using the module ena.ko .
I will check if ipxe supports such ENA adapter (I guess yes) but I will check any way.

============== lsmod =================
root@wgs ~ # modinfo ena
filename:       /lib/modules/5.10.135-gentoo-arm64/kernel/drivers/net/ethernet/amazon/ena/ena.ko
license:        GPL
description:    Elastic Network Adapter (ENA)
author:         Amazon.com, Inc. or its affiliates
alias:          pci:v00001D0Fd0000EC21sv*sd*bc*sc*i*
alias:          pci:v00001D0Fd0000EC20sv*sd*bc*sc*i*
alias:          pci:v00001D0Fd00001EC2sv*sd*bc*sc*i*
alias:          pci:v00001D0Fd00000EC2sv*sd*bc*sc*i*
alias:          pci:v00001D0Fd00000051sv*sd*bc*sc*i*
depends:        
intree:         Y
name:           ena
vermagic:       5.10.135-gentoo-arm64 SMP mod_unload aarch64
parm:           debug:Debug level (0=none,...,16=all) (int)

============== lspci ==================
root@wgs ~ # lspci -vv
00:00.0 Host bridge: Amazon.com, Inc. Device 0200
        Physical Slot: 0
        Control: I/O- Mem- BusMaster- SpecCycle- MemWINV+ VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-

00:01.0 Serial controller: Amazon.com, Inc. Device 8250 (prog-if 03 [16650])
        Physical Slot: 1
        Control: I/O- Mem+ BusMaster- SpecCycle- MemWINV+ VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Interrupt: pin B routed to IRQ 14
        Region 0: Memory at 8000c000 (32-bit, non-prefetchable) [size=4K]
        Kernel driver in use: serial

00:05.0 Ethernet controller: Amazon.com, Inc. Elastic Network Adapter (ENA)
        Subsystem: Amazon.com, Inc. Elastic Network Adapter (ENA)
        Physical Slot: 5
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV+ VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0
        Interrupt: pin A routed to IRQ 13
        Region 0: Memory at 80004000 (32-bit, non-prefetchable) [size=16K]
        Capabilities: [70] Express (v2) Root Complex Integrated Endpoint, MSI 00
                DevCap: MaxPayload 256 bytes, PhantFunc 0
                        ExtTag- RBE+ FLReset+
                DevCtl: CorrErr- NonFatalErr- FatalErr- UnsupReq-
                        RlxdOrd- ExtTag- PhantFunc- AuxPwr- NoSnoop- FLReset-
                        MaxPayload 256 bytes, MaxReadReq 128 bytes
                DevSta: CorrErr- NonFatalErr- FatalErr- UnsupReq- AuxPwr- TransPend-
                DevCap2: Completion Timeout: Not Supported, TimeoutDis- NROPrPrP- LTR-
                         10BitTagComp- 10BitTagReq- OBFF Not Supported, ExtFmt- EETLPPrefix-
                         EmergencyPowerReduction Not Supported, EmergencyPowerReductionInit-
                         FRS-
                         AtomicOpsCap: 32bit- 64bit- 128bitCAS-
                DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis- LTR- 10BitTagReq- OBFF Disabled,
                         AtomicOpsCtl: ReqEn-
        Capabilities: [b0] MSI-X: Enable+ Count=9 Masked-
                Vector table: BAR=0 offset=00002000
                PBA: BAR=0 offset=00003000
        Capabilities: [bc] Power Management version 2
                Flags: PMEClk- DSI- D1- D2- AuxCurrent=0mA PME(D0+,D1+,D2+,D3hot+,D3cold+)
                Status: D0 NoSoftRst- PME-Enable- DSel=0 DScale=0 PME-
                Bridge: PM- B3-
        Kernel driver in use: ena
        Kernel modules: ena

**Reply by stappersg** [reply]
So the device seems to have [_PCI_](https://en.wikipedia.org/wiki/Peripheral_Component_Interconnect) hardware.

To be corrected is this command line:
```text
make bin-aarch64/ipxe.foo  DEBUG=pci,efi-pci
```

I had to guess the `bin-aarch64` architecture and the `ipxe.foo` executable type.

The `pci` and `efi_pci` do match `.c`-files in the `src` directory ( `interface/efi/efi_pci.c` and `drivers/bus/pci.c` ). Effect is that debug output will be printed during run. Idea is getting more information about the _PCI_ hardware in the device.

**Comment by jorgeventura** [comment]
Hi @stappersg ,
I had a chance to be in the efi shell few days ago and when I issued the command "devices" at prompt, there was no network devices, just block devices. Does ipxe use devices presented by efi or scans devices by itself ?

Ventura

**Reply by stappersg** [reply]
To be corrected is this command line:
```text
make bin-aarch64/ipxe.foo  DEBUG=pci,efi_pci
```

Do read that as something like _"share information about what is happening at your side"_, so make it possible that others can reproduce what you have. Create common interest.

**Reply by jorgeventura** [reply]
I don't know how to do this.
I will give up.

Thank you.

**Reply by stappersg** [reply]
```text
On Mon, Sep 19, 2022 at 05:43:55AM -0700, Jorge Ventura wrote:
> I don't know how to do this.
> I will give up.


Acknowledge.

Which notes will be shared with this discussion?

Which build command did bring something usefull?

 
> Thank you.

You are welcome.

I hope that you yourself liked the discovery journey.
 

Groeten
Geert Stappers
-- 
Silence is hard to parse
```

**P.S.**
What about #748 a.k.a. #620?   Is _aarch64_ covered?

**Reply by jorgeventura** [reply]
Ok. I am back. Before I continue how do I do what you ask me to do?
I finally got list of devices presented by the EFI Shell at AWS instance and the ethernet hardware is there. My concern was that the ethernet was not available when iPXE was started but apparently this is no happening. In the AWS EFI menu; there is no option to boot by PXE or jump to "EFI Shell". To get at EFI shell I had to delete the BOOTAA64.efi file from the disk.

![Untitled](https://user-images.githubusercontent.com/85666/191111803-df7fe0ae-32d5-4575-8a98-16331cf20f20.png)

**Comment by jorgeventura** [comment]
0000:00:05.0 probe failed: Connection timed out (https://ipxe.org/4cc94092)
EFIPCI 0000:00:05.0 could not probe driver "ena-vf": Connection timed out (https://ipxe.org/4cc94092)

Now we are beginning to see something! I am going to include "ena" in DEBUG list as advised in the site.

**Comment by jorgeventura** [comment]
The message is saying something about latency time and changed to 32 but apparently was not sufficient to fix the issue.
The message about latency is coming from pci driver caused by ena driver taking too long to reset. I don't know if increasing the ENA_RESET_MAX_WAIT_MS ( 1000ms) to 5000ms can fix this problem. I will try.

iPXE initialising devices...
EFIPCI 0000:00:00.0 (1d0f:0200 class 060000) has no driver
EFIPCI 0000:00:01.0 (1d0f:8250 class 070003) has no driver
EFIPCI 0000:00:04.0 (1d0f:8061 class 010802) has no driver
EFIPCI 0000:00:05.0 (1d0f:ec20 class 020000) has driver "ena-vf"
EFIPCI 0000:00:05.0 (1d0f:ec20 class 020000) has driver "ena-vf"
EFIPCI 0000:00:05.0 (1d0f:ec20 class 020000) has driver "ena-vf"
0000:00:05.0 (1d0f:ec20) has driver "ena-vf"
0000:00:05.0 has mem 80110000 io 0 irq 4
0000:00:05.0 latency timer is unreasonably low at 0. Setting to 32.
ENA 0x785a6730 AQ [785a6e00,785a6e80) ACQ [785a6d80,785a6e00)
ENA 0x785a6730 timed out waiting for admin request 0x0:
785a6e00 : 00 00 09 01 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6e10 : 00 1c 00 00 00 70 5a 78-00 00 00 00 00 00 00 00 : .....pZx........
785a6e20 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6e30 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6d80 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6d90 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6da0 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6db0 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
ENA 0x785a6730 AQ and ACQ destroyed
0000:00:05.0 probe failed: Connection timed out (https://ipxe.org/4cc94092)
EFIPCI 0000:00:05.0 could not probe driver "ena-vf": Connection timed out (https://ipxe.org/4cc94092)
EFIPCI 0000:00:05.0 (1d0f:ec20 class 020000) has driver "ena-vf"
0000:00:05.0 (1d0f:ec20) has driver "ena-vf"
0000:00:05.0 has mem 80110000 io 0 irq 4
0000:00:05.0 latency timer is unreasonably low at 0. Setting to 32.
ENA 0x785a6730 AQ [785a6e00,785a6e80) ACQ [785a6d80,785a6e00)
ENA 0x785a6730 timed out waiting for admin request 0x0:
785a6e00 : 00 00 09 01 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6e10 : 00 1c 00 00 00 70 5a 78-00 00 00 00 00 00 00 00 : .....pZx........
785a6e20 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6e30 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6d80 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6d90 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6da0 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
785a6db0 : 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 : ................
ENA 0x785a6730 AQ and ACQ destroyed
0000:00:05.0 probe failed: Connection timed out (https://ipxe.org/4cc94092)
EFIPCI 0000:00:05.0 could not probe driver "ena-vf": Connection timed out (https://ipxe.org/4cc94092)
ok

**Comment by jorgeventura** [comment]
I changed ENA_RESET_MAX_WAIT_MS from 1000 to 5000, ENA_ADMIN_MAX_WAIT_MS from 5000 to 15000 and pci_latency from 32 to 128 and repeat the test with the same problem.

I also tried using a much faster aws instance (arm64, 8 cores, 32 Gb ram and hd ssd3) to check if it's something related to a very slow instance but nothing change.

I think now is time to stop the discussion here because I think we are facing a real issue with pci+ena. I have to spend more time diving in the ena drive.

**Reply by stappersg** [reply]
An [`ena` issue](https://github.com/ipxe/ipxe/issues/699) and repeating
* Which notes will be shared with this discussion?
* Which build command did bring something usefull?

**Reply by jorgeventura** [reply]
I was reading the issue #699 and I don't think I have something new to contribute. Something was already done but it's not working with the hardware I am using. The right thing to do in my case is to study the change and try to understand what has to be done.

Question: Is the ena drive the same from Linux with few changes to adapt to iPXE?

**Reply by stappersg** [reply]
Aim for _win/win_.

**Reply by stappersg** [reply]
Commit https://github.com/ipxe/ipxe/commit/081b3eefc48953795122c4b1a38430498e2bb2b6 fixes an `ena` issue.

**Reply by jorgeventura** [reply]
I didn't have success but the change refer to BIOS and I am using EFI. Does it matters?

**Reply by stappersg** [reply]
> Does it matter?

Yes, aiming for _win/win_ does matter!

**Reply by jorgeventura** [reply]
I was wrong when I reported no success. The commit 081b3ee is working also in AWS at least running on t4g.micro/arm64/us-east-2b .

![Untitled](https://user-images.githubusercontent.com/85666/191611327-a3911c92-4be5-482b-927e-46afbc6438ec.png)

One last question: Is it possible to include static routes so that I can use iSCSI target in a different network ? I suspect the answer is nope.
